prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -1085,34 +992,86 @@ void rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
 		rxr_pkt_entry_release_rx(ep, pkt_entry);
 		return;
 	}
-	
-	if (pkt_entry->addr == FI_ADDR_NOTAVAIL) {
-		if (pkt_type >= RXR_REQ_PKT_BEGIN && rxr_pkt_req_raw_addr(pkt_entry)) {
+
+	/*
+	 * Attempt to inject or infer peer address if not available. This usually
+	 * happens when the endpoint receives the first packet from a new peer.
+	 * There is an edge case for EFA endpoint - the device might lose the address
+	 * handle of a known peer due to a firmware bug and return FI_ADDR_NOTAVAIL.
+	 * The provider needs to infer the address by searching the reverse AV table/map,
+	 * and re-insert the address to the AV if the peer address is found.
+	 */
+	if (pkt_entry->addr == FI_ADDR_NOTAVAIL && pkt_type >= RXR_REQ_PKT_BEGIN) {
+#ifdef SUPPORT_EX_CQ
+		if (lower_ep_type == EFA_EP) {
```

Comment:
The logic is getting super complicated here. I think it would be better to have a function for efa device and a function for shm.